File: go/array/11_max_area.go
package array

// maxArea  双指针 找出最大的面积
func maxArea(height []int) int {

	l, r := 0, len(height)-1

	maxVal := 0

	for l < r {

		val := min(height[l], height[r]) * (r - l)
		if maxVal < val {
			maxVal = val
		}

		if height[l] < height[r] {
			// 如果左侧指针的长度 小于右侧
			// 左侧指针左移 有可能会加大面积
			l++
		} else {
			r--
		}

	}
	return maxVal
}

func max(x, y int) int {
	if x > y {
		return x
	}

	return y
}

File: rust/array/11_max_area.rs



pub fn max_area(height: Vec<i32>) -> i32 {

    let (mut  l, mut r) = (0 as usize, height.len()-1);
    let mut max_val = 0;

    while l < r {
        
        max_val = max(max_val, (min(height[l], height[r]) * (r - l) as i32));
        
        if height[l] < height[r] {
            l += 1 ;
        }else {
            r -= 1; 
        }

    }
    max_val
}